packages/core/contracts/chainbridge/SourceOracle.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.6.0;
+
+import "./BeaconOracle.sol";
+
+/**
+ * @title Simple implementation of the OracleInterface that is intended to be deployed on Mainnet and used
+ * to communicate price request data cross-chain with Sink Oracles on non-Mainnet networks. An Admin can publish
+ * prices to this oracle. An off-chain relayer can subsequently see when prices are published and signal to publish
+ * those prices to any non-Mainnet Sink Oracles.
+ * @dev This contract should be able to make price requests to the DVM, and the Admin capable of making and publishing
+ * price reqests should be an off-chain relayer capable of detecting signals from the non-Mainnet Sink Oracles.
+ */
+/**
+ * @title Extension of BeaconOracle that is intended to be deployed on Mainnet to give financial
+ * contracts on non-Mainnet networks the ability to trigger cross-chain price requests to the Mainnet DVM. This contract
+ * is responsible for triggering price requests originating from non-Mainnet, and broadcasting resolved price data
+ * back to those networks.
+ * @dev The intended client of this contract is some off-chain bot watching for resolved price events on the DVM. Once
+ * that bot sees a price has resolved, it can call `publishPrice()` on this contract which will call the local Bridge
+ * contract to signal to an off-chain relayer to bridge a price request to another network.
+ */
+contract SourceOracle is BeaconOracle {
```

Comment:
Do we deploy one of these per L2? Or do we deploy only one of these and it would service all other chains? I believe the latter, right? In that case (this isn't necessarily a problem), we should just be aware that someone can forward price results to any other chain if they want, and those will fail to push on the other side (because they were never requested). Not sure if this is gracefully handled by the client.

Reply:
We deploy a `Source` on Mainnet (or any other EVM that has a DVM on it) which can make price requests to and fetch price resolution data from the DVM. And we deploy one `Sink` per L2 that wants to use the `Source` as the source of truth.

> we should just be aware that someone can forward price results to any other chain if they want, and those will fail to push on the other side (because they were never requested). Not sure if this is gracefully handled by the client.

Great point. We actually configure the off-chain relayer with a `config.json` listing out all networks it should service, so its not a problem if the `destinationChainId` doesn't point to a real network:

```
{
  "chains": [
    {
      "name": "Goerli",
      "type": "ethereum",
      "id": "0",
      "endpoint": "wss://goerli-light.eth.linkpool.io/ws",
      "from": "0x",
      "opts": {
        "bridge": "0x",
        "genericHandler": "0x"
      }
    },
    {
      "name": "Rinkeby",
      "type": "ethereum",
      "id": "1",
      "endpoint": "wss://rinkeby-light.eth.linkpool.io/ws",
      "from": "0xrelayer",
      "opts": {
        "bridge": "0x",
        "genericHandler": "0x"
      }
    }
  ]
}
```

Reply:
What if we had 3 networks, two L2s (id 2 and 3) and 1 L1 (id 1)? What would happen if id 2 requests a price, and someone pushes the price to id 2 and id 3? What would happen when the bridge tries to push the price to id 3?

Also, what would happen if someone tried to publish it to id 1 (the same as the source chain)?

Reply:
> What would happen if id 2 requests a price, and someone pushes the price to id 2 and id 3? What would happen when the bridge tries to push the price to id 3?

The `publishPrice` on the id 3 `SinkOracle` would revert because there was no request pending.

> Also, what would happen if someone tried to publish it to id 1 (the same as the source chain)?

This has to happen on L1 to start the cross-chain communication. Someone can call `publishPrice` on L1 to trigger a `Bridge.deposit` call on L1. No one can directly call `publishPrice` on L2 except for the `GenericHandler` which would be called by `Bridge.deposit` on L2.

Do these answers make sense and do you agree with them?